Sources/LanguageServerProtocol/SupportTypes/RegistrationOptions.swift

```diff
@@ -12,12 +12,23 @@
 
 import Foundation
 
-/// An event describing a file change.
-public protocol RegistrationOptions: Codable, LSPAnyCodable, Hashable {
+/// Protocol for capability registration options, which must be encodable to
+/// `LSPAny` so they can be included in a `Registration`.
+public protocol RegistrationOptions: Hashable {
+  func encodeIntoLSPAny(dict: inout [String: LSPAny])
+}
+
+fileprivate func encode(strings: [String]) -> LSPAny {
+  var values = [LSPAny]()
+  values.reserveCapacity(strings.count)
+  for str in strings {
+    values.append(.string(str))
+  }
+  return .array(values)
 }
 
 /// General text document registration options.
-public struct TextDocumentRegistrationOptions: RegistrationOptions {
+public class TextDocumentRegistrationOptions: RegistrationOptions, Hashable {
```

Comment:
I think making this a class is bad idea.  The most interesting functionality this type provides is serialization, but that creates a foot gun for types that inherit from it, since they need to remember to override the serialization and if they try to call `super.encodeIntoLSPAny(...)` they need to know how the superclass serialization is implemented in order to interop with it correctly.

Also, about half the code in this type and in the types that inherit from it is now from Hashable/Equatable implementations that would be free for a struct.

Finally, it's *surprising* this is a class since none of our other simple data types like this have reference semantics.

If what we need is a type that can be used in a dictionary, I suggest making a protocol for what you need.

Reply:
The main point of this type was to make it easier to introduce types that extend `TextDocumentRegistrationOptions`, inheritance it simplest way to do this in Swift but I can also instead make the sub-types store a `TextDocumentRegistrationOptions` and manually call the methods. I'm not sure how a protocol would help with that. WDYT?

Reply:
SGTM.  My comment about the protocol was due to you using `[Registration: TextDocumentRegistrationOptions]`, but if you don't need dynamic behaviour there great.

Reply:
Ah I see, yeah a protocol would be useful for that but for now we don't need it.

Reply:
Actually, going with a protocol since it will be useful for semantic highlighting

Reply:
Done